### ISSUE TITLE
Using ArrayImageSource With Mean/Volume Estimator Crashes

### ISSUE THREAD

**Issue by garrettwrong** [ComputationalCryoEM/ASPIRE-Python#426]
Found when trying to use researcher provided MATLAB reference data for #406 .

Essentially, `ArrayImageSource` does not correctly set attributes relating to rotations (`rots` property, and/ `_rotations`) required in some of the volume estimation code.  At least that is the first issue... Probably these just need to be initialized to zero equivalents. 

Reproduced in `develop`. The following isolated bit of code extends the unit tests to cover this case:

```
+++ b/tests/test_image.py
@@ -4,7 +4,9 @@ from unittest import TestCase
 import numpy as np
 from scipy import misc
 
+from aspire.basis import FBBasis3D
 from aspire.image import Image, _im_translate2
+from aspire.reconstruction import MeanEstimator
 from aspire.source import ArrayImageSource
 
 DATA_DIR = os.path.join(os.path.dirname(__file__), "saved_test_data")
@@ -12,13 +14,14 @@ DATA_DIR = os.path.join(os.path.dirname(__file__), "saved_test_data")
 
 class ImageTestCase(TestCase):
     def setUp(self):
+        self.dtype = np.float64
         # numpy array for top-level functions that directly expect it
-        self.im_np = misc.face(gray=True).astype(np.float64)[np.newaxis, :768, :768]
+        self.im_np = misc.face(gray=True).astype(self.dtype)[np.newaxis, :768, :768]
         # Independent Image object for testing Image methods
-        self.im = Image(misc.face(gray=True).astype(np.float64)[:768, :768])
+        self.im = Image(misc.face(gray=True).astype(self.dtype)[:768, :768])
         # Construct a simple stack of Images
         self.n = 3
-        self.ims_np = np.empty((3, *self.im_np.shape[1:]), dtype=self.im_np.dtype)
+        self.ims_np = np.empty((3, *self.im_np.shape[1:]), dtype=self.dtype)
         for i in range(self.n):
             self.ims_np[i] = self.im_np * (i + 1) / float(self.n)
         # Independent Image stack object for testing Image methods
@@ -53,6 +56,12 @@ class ImageTestCase(TestCase):
         im = src.images(start=0, num=np.inf)
         self.assertTrue(np.allclose(im.asnumpy(), self.im_np))
 
+    def testArrayImageSourceVol(self):
+        src = ArrayImageSource(self.im)
+        basis = FBBasis3D((8, 8, 8), dtype=self.dtype)
+        estimator = MeanEstimator(src, basis, preconditioner="none")
+        _ = estimator.estimate()
+
     def testImageSqrt(self):
         self.assertTrue(np.allclose(self.im.sqrt().asnumpy(), np.sqrt(self.im_np)))
 ```
 
 tb:
 
 ```
 ======================================================= FAILURES ========================================================
_________________________________________ ImageTestCase.testArrayImageSourceVol _________________________________________

self = <tests.test_image.ImageTestCase testMethod=testArrayImageSourceVol>

    def testArrayImageSourceVol(self):
        src = ArrayImageSource(self.im)
        basis = FBBasis3D((8, 8, 8), dtype=self.dtype)
        estimator = MeanEstimator(src, basis, preconditioner="none")
>       _ = estimator.estimate()

tests/test_image.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/aspire/reconstruction/estimator.py:68: in estimate
    b_coeff = self.src_backward()
src/aspire/reconstruction/estimator.py:85: in src_backward
    batch_mean_b = self.src.im_backward(im, i) / self.n
src/aspire/source/image.py:509: in im_backward
    vol = im.backproject(self.rots[start : start + num, :, :])[0]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <aspire.source.image.ArrayImageSource object at 0x11cea6160>

    @property
    def rots(self):
        """
        :return: Rotation matrices as a n x 3 x 3 array
        """
>       return self._rotations.as_matrix().astype(self.dtype)
E       AttributeError: 'ArrayImageSource' object has no attribute '_rotations'

src/aspire/source/image.py:195: AttributeError
 ```

**Comment by janden** [comment]
Mean estimation is not defined if rotations are present, so the solution here should be to raise some sort of error.

**Comment by garrettwrong** [comment]
Rotations are not present. We're just trying to pass a static array of images through...

**Comment by garrettwrong** [comment]
Basically, if we try to use most of ASPIRE with anything other than `Simulation` I have zero expectation it will work.

**Comment by janden** [comment]
Sorry I mistyped. Mean is not defined if rotations *are not* present. This is expected behavior. It shouldn't work.

**Comment by garrettwrong** [comment]
Thanks for trying to help, but I think we're working two different angles. I do not agree with you. Here's what I'm thinking:

1) I actually have rotations available, but the source does not admit them.. that is a different problem that comes later. I was intentionally avoiding since it feels mostly irrelevant at this point. I actually care little about the algorithm at this moment.  I simply have some junk in an array and I should be able to push it through two different ways and see the exact same output (whatever it is, crash, other junk, 42 etc).
2) I can disable all the knobs in `Simulation` (to basically get a stack of id images, confirmed) and the code in question _does not crash_ (you say it should... :see_no_evil: ... maybe it should, but again, different issue to why de-doped Simulation vs ArrayImageSource of that array do not yield the same exact thing... either both crash identically, or both yield the same junk right).

Pushing the exact same inputs through seems to fail in a variety of meta code reasons (attributes, filters, filter nan values). Not because an algorithm. This is the stuff this issue should fix, at least for this case. There will be others no doubt.

What you are saying is almost certainly correct, but you are thinking about a totally different concern than me. We're not there yet... 

I feel we need a way to take provided data in an array and pass to components in ASPIRE. ArrayImageSource is supposed to be that way. ArrayImageSource does not seem to function. Every time we try to use it with something new we find similar problems. I would bet real money you will find similar problems with the other sources too. ASPIRE-Python code from the last two years or so is baked expecting logic that only exists pretty much exclusively in `Simulation`. I can't fix them all at once, but I think we can fix on them a bit here and there to at least make sure the right attributes are populated and arguments that might be required are exposed as we work with specific features (in this case MeanEstimator).

If you do not agree I'd like to understand more about it.  Antithetically, if `ArrayImageSource` is not expected to function for main components, we instead need to document this and raise it specifically.... We would also lose most hope of easily validating things since `Simulation` is at heart a complicated noise machine. Validation using some provided external data then might require constantly custom editing source code (like the MATLAB codes generally did). Alternatively we force exporting ASPIRE-Python simulation data and importing into other arenas (with no users, this definitely a non starter).

**Comment by janden** [comment]
I'm not following. If you want to estimate the mean, you need a few things: images, filters, rotations, amplitudes, and shifts. Those are all required by the mean reconstruction algorithm. Just supplying images will not work unless you also add these other things. I agree that `ArrayImageSource` should provide a means for doing this (point 1 above, if I understood correctly).

For point 2, I'm not sure what you mean by disabling things in `Simulation`. Are you removing the rotation attributes? Then mean estimation should no longer work. I'm not sure why it is working for you. I don't see how 1 and 2 are the same inputs.

And yes, `ArrayImageSource` has been there mostly as a stub for supplying images that do not expect any metadata (like the preprocessing). If we want it to work with some of the heavier algorithms, it will need to be modified to supply that metadata (not just rotations, but filters, shifts, etc.). This would be useful. So far, the algorithms have been targeting `Simulation` and `RelionSource` – simulated and external data on disk. That has been the priority.

**Comment by janden** [comment]
Ok I see now what you're doing for 2. You're supplying zero angles, which should correspond to identity rotation matrices, which is what you set the rotations to in `ArrayImageSource`. While this may not give an error directly in the mean estimation, CG won't converge because the inversion is ill-posed. Is that what you're seeing?

**Comment by janden** [comment]
Looks like the it was the missing filters, so `src.eval_filter_grid` was just uninitialized data. Most of the time this means at least one nan which contaminates the whole kernel computation. This seems to work on my end:
```
diff --git a/src/aspire/source/image.py b/src/aspire/source/image.py
index ae52b05..0a56e70 100644
--- a/src/aspire/source/image.py
+++ b/src/aspire/source/image.py
@@ -764,7 +764,8 @@ class ArrayImageSource(ImageSource):
         self._cached_im = im
 
         # Create filter indices, engages setter of things we don't need but still helps the code crash a lot.
-        self.filter_indices = None
+        self.unique_filters = [IdentityFilter()]
+        self.filter_indices = np.zeros(im.n_images)
 
     def _rots(self):
         """
```
Again, the inversion should still not work due to the ill-posedness, but it the nans should be gone.

**Comment by garrettwrong** [comment]
I tried the IdentityFilter (and also changing the default nan value in filter value setting).  It did not occur to me at all to zero the indices. Nice. Progress. Thanks a lot!

So that does give identical results.  Now they both converge. (heh, yeah I know.. we will have to get to that 🙉  ).

I think now I feel comfortable plumbing through some optional rotations via the `ArrayImageSource` instantiator.  Probably it makes sense to have the getter give a nice message if some (heavy) consumer code asks for an attribute like this and it is `None`. (Instead of hacking in the identity like it did for testing). That will let me pass the image array and the rotation array supplied by a researcher. It will also let me write a unit test where I have total (and very clear) control over the rotations (can get from Simulation) to check that it is identical to a Simulation source... (for now I pray Simulation is correct then...)

I hope my way of approaching this makes more sense.

**Comment by janden** [comment]
> So that does give identical results. Now they both converge. (heh, yeah I know.. we will have to get to that hear_no_evil ).

Yes well… Different issue.

> I think now I feel comfortable plumbing through some optional rotations via the `ArrayImageSource` instantiator. Probably it makes sense to have the getter give a nice message if some (heavy) consumer code asks for an attribute like this and it is `None`.

Yes this makes sense to me. Rotations is the only tricky one since it lacks a sensible default. For filters, it should default to a single `IdentityFilter` and all filter indices pointing to that one. For offsets, they should all be zero, and for amplitudes, they should all be ones, but there is no default rotation. So the sensible thing to do might be to default to None, allow optional rotations to go through with the constructor, and raise some sort of error someone asks for rotations and they're not present.

Now when this works for `ArrayImageSource`, we should see how much of this can be propagated up to `ImageSource`. While `Simulation` should have all of this metadata (should, yes), that might not be the case for others, such as `RelionSource`. To avoid seeing similar hard-to-find bugs there, it might make sense to implement a lot of this at the top level so it gets inherited by all the sources and overridden as needed. Does that make sense?

**Comment by garrettwrong** [comment]
I see the problem as business logic that is somewhat intimate is split in both `ImageSource` and `Simulation`.

I want to believe it will be the only tricky one 🙂 .  I suspect you may want to be able to override all of them at some point.

I didn't catch this message because I was working on the patch, but I think what I came up with is mostly in line with what you suggest. Take a peek at #427 . I think the unit test should cover what we were discussing today....

**Comment by garrettwrong** [comment]
Closed by #427
